dev/src/reference.ts

```diff
@@ -398,12 +402,18 @@ export class DocumentReference<T = DocumentData> implements Serializable {
    *   console.log(`Document written at ${res.updateTime}`);
    * });
    */
-  set(data: T, options?: SetOptions): Promise<WriteResult> {
-    const writeBatch = new WriteBatch(this._firestore);
-    return writeBatch
-      .set(this, data, options)
-      .commit()
-      .then(([writeResult]) => writeResult);
+  set(data: T | Partial<T>, options?: SetOptions) {
+    let writeBatch = new WriteBatch(this._firestore);
+    if (
+      options &&
+      (options.merge || options.mergeFields) &&
+      !isDefaultConverter(this._converter)
+    ) {
+      writeBatch = writeBatch.set(this, data as Partial<T>, options);
+    } else {
+      writeBatch = writeBatch.set(this, data as T, options);
```

Comment:
If you add "the catchall" signature as a overload to WriteBatch you won't have to do this dance:

Reply:
wow that's clean! thanks